app/Foliage/CmdBuild.hs

```diff
@@ -272,11 +266,28 @@ getPackageVersions inputDir = do
 
   forP metaFiles $ preparePackageVersion inputDir
 
-prepareIndexPkgCabal :: PackageId -> UTCTime -> FilePath -> Action Tar.Entry
-prepareIndexPkgCabal pkgId timestamp filePath = do
-  need [filePath]
-  contents <- liftIO $ BS.readFile filePath
-  pure $ mkTarEntry (BL.fromStrict contents) (IndexPkgCabal pkgId) timestamp
+prepareIndexPkgCabal :: PackageId -> UTCTime -> FilePath -> [(UTCTime, FilePath)] -> Action [Tar.Entry]
+prepareIndexPkgCabal pkgId timestamp originalFilePath patches = do
+  need (originalFilePath : map snd patches)
+  original <- liftIO (BL.readFile originalFilePath)
+  patchesApplied <- applyRevisionsInOrder [(timestamp, original)] patches
+  pure $ map (\(time, content) -> mkTarEntry content (IndexPkgCabal pkgId) time) patchesApplied
+
+applyRevisionsInOrder :: [(UTCTime, BL.ByteString)] -> [(UTCTime, FilePath)] -> Action [(UTCTime, BL.ByteString)]
+applyRevisionsInOrder acc [] = pure acc
+applyRevisionsInOrder acc (patch@(revisionTime, _) : remainingPatches) = do
+  newContent <- applyRevision (snd $ last acc) patch
+  applyRevisionsInOrder (acc ++ [(revisionTime, newContent)]) remainingPatches
```

Comment:
Appending to the end of a list that you're constructing as you go is quadratic and generally a bad idea. Either cons on the front and reverse at the end or use a DList